libraries/AP_ExternalAHRS/AP_ExternalAHRS_VectorNav.cpp

```diff
@@ -200,22 +280,165 @@ bool AP_ExternalAHRS_VectorNav::check_uart()
+            request_sent = now;
+        }
+
+        int16_t nbytes = uart->available();
+        while (nbytes-- > 0) {
+            char c = uart->read();
+            if (decode(c)) {
+                return;
+            }
+        }
+    }
+}
+
+// add a single character to the buffer and attempt to decode
+// returns true if a complete sentence was successfully decoded
+bool AP_ExternalAHRS_VectorNav::decode(char c)
+{
+    switch (c) {
+    case ',':
+        // end of a term, add to checksum
+        nmea.checksum ^= c;
+        FALLTHROUGH;
+    case '\r':
+    case '\n':
+    case '*':
+    {
+        if (nmea.sentence_done) {
+            return false;
+        }
+        if (nmea.term_is_checksum) {
+            nmea.sentence_done = true;
+            uint8_t checksum = 16 * char_to_hex(nmea.term[0]) + char_to_hex(nmea.term[1]);
+            return ((checksum == nmea.checksum) && nmea.sentence_valid);
+        }
+
+        // null terminate and decode latest term
+        nmea.term[nmea.term_offset] = 0;
+        if (nmea.sentence_valid) {
+            nmea.sentence_valid = decode_latest_term();
+        }
+
+        // move onto next term
+        nmea.term_number++;
+        nmea.term_offset = 0;
+        nmea.term_is_checksum = (c == '*');
+        return false;
+    }
+
+    case '$': // sentence begin
+        nmea.sentence_valid = true;
+        nmea.term_number = 0;
+        nmea.term_offset = 0;
+        nmea.checksum = 0;
+        nmea.term_is_checksum = false;
+        nmea.sentence_done = false;
+        return false;
+    }
+
+    // ordinary characters are added to term
+    if (nmea.term_offset < sizeof(nmea.term) - 1) {
+        nmea.term[nmea.term_offset++] = c;
+    }
+    if (!nmea.term_is_checksum) {
+        nmea.checksum ^= c;
+    }
+
+    return false;
+}
+
+// decode the most recently consumed term
+// returns true if new sentence has just passed checksum test and is validated
+bool AP_ExternalAHRS_VectorNav::decode_latest_term()
+{
+    switch (nmea.term_number) {
+        case 0:
+            if (strcmp(nmea.term, "VNRRG") != 0) {
+                return false;
+            }
+            break;
+
+        case 1:
+            if (nmea.register_number != strtoul(nmea.term, nullptr, 10)) {
+                return false;
+            }
+            break;
+
+        case 2:
+            strncpy(model_name, nmea.term, sizeof(model_name));
+            break;
+
+        default:
+            return false;
+    }
+    return true;
+}
+
 void AP_ExternalAHRS_VectorNav::update_thread()
 {
-    if (!port_opened) {
-        // open port in the thread
-        port_opened = true;
-        uart->begin(baudrate, 1024, 512);
-        send_config();
+    // Open port in the thread
+    uart->begin(baudrate, 1024, 512);
+
+    // Reset and wait for module to reboot
+    // VN_100 takes 1.25 seconds
+    //nmea_printf(uart, "$VNRST");
+    //hal.scheduler->delay(3000);
```

Comment:
Had to comment this as it breaks baro calibration if it comes up too slowly. Needs a fair bit of baro rework to allow "late" calibrations. Once that is done it would be nice to un-comment this again.